Refactor NavigateToPose action client into rviz Nav2 panel and expose Cancel/Shutdown

bpwilcox opened the pull request:
## Description of contribution
- creates `GoalPoseUpdater` QObject to emit signal for goal poses. Creates a global object to be accessed in both the nav2 panel and nav2 goal tool
- modifies`nav2_rviz_plugin::GoalTool ` to send goal poses via Qt signal using `GoalPoseUpdater' QObject
- adds action client and supporting goal-related functionality to `Nav2Panel`
- the cancel button state has been added to the panel
- the shutdown button now transitions to cancel when the NavigateToPose action is active, when the action is over or is canceled by the user, the shutdown button state is again present
  - created custom  `ROSActionEvent` and `ROSActionTransition` to  handle Qt button state transitions 

![Screenshot from 2019-07-17 14-57-54](https://user-images.githubusercontent.com/22353511/61415174-ddc09700-a8a4-11e9-9c57-81fbe09c6e78.png)

mjeronimo commented on nav2_rviz_plugins/src/nav2_panel.cpp:
```diff
@@ -48,29 +54,101 @@ Nav2Panel::Nav2Panel(QWidget * parent)
+  // State entered while the NavigateToPose action is active
+  running_ = new QState();
+  running_->setObjectName("running");
+  running_->assignProperty(start_stop_button_, "text", "Cancel");
```
The start_stop_button is not longer just start and stop, but now also cancel. Perhaps rename the variable?

bpwilcox replied:
Any name suggestions? You can argue cancel is semantically a stop as well, if you don't mind the term overloading.

mjeronimo replied:
Yeah, that's true. Never mind on the original comment. I think it's clear enough as is.

SteveMacenski commented on nav2_rviz_plugins/include/nav2_rviz_plugins/goal_tool.hpp:
```diff
@@ -50,8 +51,12 @@ class RVIZ_DEFAULT_PLUGINS_PUBLIC GoalTool : public rviz_default_plugins::tools:
 private:
-  std::unique_ptr<NavigationDialog> navigation_dialog_;
+  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr publisher_;
+  std::unique_ptr<rviz_common::properties::StringProperty> topic_property_;
```
Why does this need to be a pointer?

bpwilcox replied:
I was just carrying over this code form the InitalPoseTool in from `rviz_default_plugins`.

SteveMacenski replied:
OK, far be it for me to argue with the rviz folks

SteveMacenski commented on nav2_rviz_plugins/include/nav2_rviz_plugins/nav2_panel.hpp:
```diff
@@ -43,9 +87,40 @@ class Nav2Panel : public rviz_common::Panel
+  rclcpp::Node::SharedPtr goal_node_;
+
+  // Executor needed to spin node in thread
+  std::unique_ptr<rclcpp::executors::SingleThreadedExecutor> executor_;
```
Similar question: why does this need to be a ptr?

bpwilcox commented:
Addressed feedback. I decided to remove the `nav2_rviz_plugins::GoalTool` because I realized it was going to be no different that the existing `rviz_default_plugins::GoalTool` since the panel subscribes to the goal topic.

crdelsey commented:
@bpwilcox Please change the topic name or something to prevent both the panel and bt_navigator from responding to the topic.

bpwilcox commented:
I've updated this PR to address feedback from our meeting to not use a subscriber in the nav2 panel to listen for the goal pose. Per @crdelsey's suggestion, I defined a custom global object to allow access between the panel and tool plugins in rviz. This custom `QObject` allows us to send the goal pose from the goal tool to the panel by emitting a Qsignal with the updated pose.

bpwilcox commented:
@mjeronimo  @crdelsey  Can you re-review since the changes I made?

crdelsey commented on nav2_rviz_plugins/src/nav2_panel.cpp:
```diff
@@ -48,29 +58,99 @@ Nav2Panel::Nav2Panel(QWidget * parent)
+
+  auto options = rclcpp::NodeOptions().arguments(
+    {"__node:=navigation_dialog_action_client"});
+  client_node_ = std::make_shared<rclcpp::Node>("_", options);
```
What are you doing here? Why create the node by the name of "_" but then use a argument to rename it?

bpwilcox replied:
This wasn't my code but moved from the `navigation_dialog.cpp` before my refactor. Seemingly, this is done several places in the code (see `bt_navigator.cpp`). If I recall, this was done to support remapping?

crdelsey commented on nav2_rviz_plugins/src/nav2_panel.cpp:
```diff
@@ -93,6 +173,108 @@ Nav2Panel::onShutdown()
+  if (status == action_msgs::msg::GoalStatus::STATUS_ACCEPTED ||
+    status == action_msgs::msg::GoalStatus::STATUS_EXECUTING)
+  {
+    state_machine_.postEvent(new ROSActionQEvent(true));
```
If I understand this correctly, you are posting a bool as an event. Is that correct? That conveys little information. The events should be an enum at a minimum. That way you could post something like "NAV_ACTIVATED"  or "NAV_DEACTIVATED"

bpwilcox replied:
Same functionality but I suppose an enum indicates it is more specific to actions being active/inactive.

crdelsey commented on nav2_rviz_plugins/src/nav2_panel.cpp:
```diff
@@ -93,6 +173,108 @@ Nav2Panel::onShutdown()
+  // Send the goal pose
+  goal_.pose = pose;
+
+  // Enable result awareness by providing an empty lambda function
```
What's "result awareness"? Can you elaborate on this comment

bpwilcox replied:
I'm not entirely sure as this wasn't my code/comment, just part of refactor, perhaps @mjeronimo could explain better.

mjeronimo replied:
"Result awareness" was introduced into the ROS2 action client. There were some relatively recent changes (a couple months ago or so), where one now provides callbacks in the SendGoalOptions for the goal response, feedback, and result. The SendGoalOptions are then passed to the client's async_send_goal method. In the navigation stack we already had code written to the old style of explicitly sending the goal request, etc. This SendGoalOptions change broke our code. One way I found to work around it was to provide an empty callback for the result_callback. Providing this callback causes the ROS2 to set the result aware flag (whatever that means) and allows us to call the existing functions. Alternatively, we could update our client code to conform to the new callback model. See the ROS2 code in rclcpp_action for more about "result awareness." I haven't found any docs.

bpwilcox commented:
Here's a state diagram for the transitions of the rviz panel button.

![Rviz Panel state diagram](https://user-images.githubusercontent.com/22353511/62730354-a2f9dc80-b9d4-11e9-909b-912cd502c704.png)

bpwilcox commented on nav2_rviz_plugins/src/nav2_panel.cpp:
```diff
@@ -48,29 +58,99 @@ Nav2Panel::Nav2Panel(QWidget * parent)
 
   initial_->addTransition(start_stop_button_, SIGNAL(clicked()), starting_);
   starting_->addTransition(start_stop_button_, SIGNAL(clicked()), stopping_);
+  running_->addTransition(start_stop_button_, SIGNAL(clicked()), canceled_);
```
Done, see comment

bpwilcox commented:
Here's a class diagram for the nav2 rviz plugins:

![Rviz Panel Class Diagrams](https://user-images.githubusercontent.com/22353511/62734707-7ea2fd80-b9de-11e9-8ec1-d0d30c3c94b0.png)

So now,  the sequence from sending a goal position becomes:

![image](https://user-images.githubusercontent.com/22353511/62734823-bc078b00-b9de-11e9-8230-60cd5107bd4f.png)

mjeronimo commented on nav2_rviz_plugins/src/nav2_panel.cpp:
```diff
@@ -20,12 +20,21 @@
 #include <QPushButton>
 #include <QtConcurrent/QtConcurrent>
 #include <QVBoxLayout>
+#include <QEventTransition>
```
Some of these headers are no longer needed and can be removed (dirent, QHBoxLayout, QLabel, QEventTransition).

mjeronimo commented on nav2_rviz_plugins/src/nav2_panel.cpp:
```diff
@@ -48,29 +58,99 @@ Nav2Panel::Nav2Panel(QWidget * parent)
+    this, SLOT(onNewGoal(double,double,double,QString)));  // NOLINT
+
+  // Launch a thread to run the node
+  thread_ = std::make_unique<std::thread>(
```
You could use a QBasicTimer (see the old code) instead of creating a node just to create a timer.

mjeronimo commented on nav2_rviz_plugins/src/nav2_panel.cpp:
```diff
@@ -48,29 +58,99 @@ Nav2Panel::Nav2Panel(QWidget * parent)
+      executor_.remove_node(node->get_node_base_interface());
+    }, timer_node_);
+
+  timer_ = timer_node_->create_wall_timer(1s, std::bind(&Nav2Panel::timerActionEvent, this));
```
You could check more frequently than once per second. This would make it more responsive to detecting the completed navigation at little expense. For example, running the callback at 5Hz should be negligible.
